cluster/swarm/cluster.go

```diff
@@ -359,9 +359,13 @@ func (c *Cluster) removeEngine(addr string) bool {
 func (c *Cluster) monitorDiscovery(ch <-chan discovery.Entries, errCh <-chan error) {
 	// Watch changes on the discovery channel.
 	currentEntries := discovery.Entries{}
-	for {
+	for ch != nil || errCh != nil {
 		select {
-		case entries := <-ch:
+		case entries, ok := <-ch:
+			if !ok {
+				ch = nil
```

Comment:
I like the change of breaking the infinite loop. But I don't think we should set channel to `nil`. It has side effects.